src/main/java/io/cryostat/core/agent/LocalProbeTemplateService.java

```diff
@@ -0,0 +1,157 @@
+
+    public static final String TEMPLATE_PATH = "CRYOSTAT_PROBE_TEMPLATE_PATH";
+
+    private final FileSystem fs;
+    private final Environment env;
+
+    public LocalProbeTemplateService(FileSystem fs, Environment env) throws IOException {
+        this.fs = fs;
+        this.env = env;
+        // Sanity check
+        if (!env.hasEnv(TEMPLATE_PATH)) {
+            throw new IOException(
+                    String.format(
+                            "Probe template directory does not exist, must be set using environment variable %s",
+                            TEMPLATE_PATH));
+        }
+        // Sanity check the directory is set up correctly
+        Path probeTemplateDirectory = fs.pathOf(env.getEnv(TEMPLATE_PATH));
+        if (!fs.exists(probeTemplateDirectory)
+                || !fs.isDirectory(probeTemplateDirectory)
+                || !fs.isReadable(probeTemplateDirectory)
+                || !fs.isWritable(probeTemplateDirectory)) {
+            throw new IOException(
+                    String.format(
+                            "Probe template directory %s does not exist, is not a directory, or has incorrect permissions.",
+                            probeTemplateDirectory.toString()));
+        }
+    }
+
+    public void addTemplate(InputStream inputStream, String filename) throws Exception {
+        try (inputStream) {
+            ProbeTemplate template = new ProbeTemplate();
+            // If validation fails this will throw a ProbeValidationException with details
+            template.deserialize(inputStream);
+            Path path = fs.pathOf(env.getEnv(TEMPLATE_PATH), filename);
+            if (fs.exists(path)) {
+                throw new Exception(
+                        String.format(
+                                "Event template \"%s\" already exists", template.getFileName()));
+            }
+            fs.writeString(
+                    path,
+                    new BufferedReader(new InputStreamReader(inputStream, StandardCharsets.UTF_8))
+                            .lines()
+                            .collect(Collectors.joining("\n")));
+        } catch (Exception e) {
+            // rethrow for http handler
+            throw e;
+        }
+    }
+
+    public void deleteTemplate(String templateName) throws Exception {
+        if (!fs.deleteIfExists(fs.pathOf(env.getEnv(TEMPLATE_PATH), templateName))) {
+            throw new IOException(
+                    String.format("Probe template \"%s\" does not exist", templateName));
+        }
+    }
+
+    public String getTemplate(String templateName) throws Exception {
+        Path probeTemplatePath = fs.pathOf(env.getEnv(TEMPLATE_PATH), templateName);
+        ProbeTemplate template = new ProbeTemplate();
+        template.deserialize(fs.newInputStream(probeTemplatePath));
+        return template.serialize();
+    }
+
+    protected List<Path> getLocalTemplates() throws IOException {
+        if (!env.hasEnv(TEMPLATE_PATH)) {
```

Comment:
I think this check can be removed, since it's already checked in the constructor.

Reply:
The other check later down confirming that the directory still exists and is readable is fine, since it's possible that the directory is on a mounted volume and could go away at runtime. The `listDirectoryChildren` call should fail with some IOException in that case anyway, but it might be more evident what the exact problem was with the explicit check. But the env var value can't change during the lifetime of the process, so no need to re-check that.